gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/Diagnostics.java

```diff
@@ -0,0 +1,215 @@
+                                .map(artifact -> {
+                                    ComponentIdentifier id = artifact.getId().getComponentIdentifier();
+                                    if (id instanceof ProjectComponentIdentifier) {
+                                        Project dependencyProject =
+                                                current.project(((ProjectComponentIdentifier) id).getProjectPath());
+                                        return maybeGetSourceFileFromLocalProject(dependencyProject);
+                                    } else {
+                                        return maybeExtractFromJar(artifact.getFile(), id);
+                                    }
+                                }))
+                .filter(Optional::isPresent)
+                .flatMap(isPresent -> isPresent.get().get().stream())
+                .distinct() // would be kinda weird if multiple jars claim to provide the same diagnostic type??
+                .sorted(Comparator.comparing(value -> value.type().toString()))
+                .collect(Collectors.toList());
+
+        return SupportedDiagnostics.of(list);
+    }
+
+    @Value.Immutable
+    interface SupportedDiagnostics {
+        @Value.Parameter
+        @JsonValue
+        List<SupportedDiagnostic> get();
+
+        @JsonCreator
+        static SupportedDiagnostics of(List<SupportedDiagnostic> items) {
+            return ImmutableSupportedDiagnostics.of(items);
+        }
+    }
+
+    // This is the format the sls-spec wants list items to be. <code>{type: "foo.v1"}</code>.
+    @Value.Immutable
+    @JsonDeserialize(as = ImmutableSupportedDiagnostic.class)
+    interface SupportedDiagnostic {
+        @Value.Parameter
+        DiagnosticType type();
+    }
+
+    private static Optional<SupportedDiagnostics> maybeGetSourceFileFromLocalProject(Project proj) {
+        JavaPluginConvention javaPlugin = proj.getConvention().findPlugin(JavaPluginConvention.class);
+        if (javaPlugin == null) {
+            return Optional.empty();
+        }
+
+        Set<File> sourceFiles = javaPlugin
+                .getSourceSets()
+                .getByName("main")
+                .getResources()
+                .getAsFileTree()
+                .filter(file -> {
+                    return file.toString().endsWith(PATH_IN_JAR);
+                })
+                .getFiles();
+        if (sourceFiles.isEmpty()) {
+            return Optional.empty();
+        }
+        if (sourceFiles.size() > 1) {
+            throw new GradleException("Expecting to find 0 or 1 files, found: " + sourceFiles);
+        }
+        File file = Iterables.getOnlyElement(sourceFiles);
+        String string = null;
+        try {
+            string = new String(java.nio.file.Files.readAllBytes(file.toPath()), StandardCharsets.UTF_8).trim();
+            SupportedDiagnostics value = CreateManifestTask.jsonMapper.readValue(file, SupportedDiagnostics.class);
```

Comment:
double read is wonky, but the failure message is nice.

Reply:
Ah yep I meant to plumb the string in and read once